### PR TITLE
Android driver refactor 1.2

### PR DESCRIPTION
**Pull request by d4vidi**
- [ ] This is a small change 
- [ ] This change has been discussed in issue #<?> and the solution has been agreed upon with maintainers.

---

**Description:**
Following work on #1192 -- need to break some of the AndroidDriver monolith into sub-units. This is an official refactor finalization step - 1.2, following #2150, #2146, #2148 and #2149: Reduce instrumentation management in driver itself to bare minimum (logic extracted to new class `MonitoredInstrumentation`, which wraps around the pure `Instrumentation` class introduced at an earlier refactor point).

### CODE REVIEW

**Comment by noomorph** [comment]
Okay, I'll review tomorrow, I suppose.

**Comment by d4vidi** [comment]
@noomorph ❤️  forgot to mention - this is because I somewhat upgraded the `Deferred` class. Need your approval

**Comment by noomorph** [comment]
@d4vidi, the changes to Deferred look fine to me.

File: detox/src/devices/drivers/android/AndroidDriver.js [hunk 1]
```diff
@@ -132,15 +102,9 @@ class AndroidDriver extends DeviceDriverBase {
...
   async waitUntilReady() {
       try {
-        await Promise.race([
-          super.waitUntilReady(),
-          new Promise((resolve, reject) => {
-            this.instrumentationCloseListener = () => reject(this._getInstrumentationCrashError());
-            !this.instrumentation.isRunning() && this.instrumentationCloseListener();
-          }),
-        ]);
+        await Promise.race([super.waitUntilReady(), this.instrumentation.waitForCrash()]);
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/AndroidDriver.js:105]: This is the core of the change, in a nutshell. Very proud of this concise result of the new way to implement the same logic.

File: detox/src/devices/drivers/android/AndroidDriver.js [hunk 2]
```diff
@@ -256,56 +204,37 @@ class AndroidDriver extends DeviceDriverBase {
...
 
-    for (let attempts = 5; attempts > 0; attempts--) {
-      const pid = await this.adb.pidof(deviceId, bundleId);
-
-      if (pid > 0) {
-        return pid;
-      }
-
-      await sleep(1000);
-    }
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/AndroidDriver.js:306]: unified with the existing `retry()` util function though a new arg: `backoff: 'none'` (default is `linear`).

**Comment by d4vidi** [comment]
@noomorph any idea what to make of coverall's report? It marks new lines as uncovered - which is fine in the sense that they are new, but nevertheless many of the reported lines _are_ covered.

**Pull request merged**